### PR TITLE
Port regionWorker, step 2

### PR DESCRIPTION
**Pull request by xiaobogaga**
@coocood PTAL, thanks.

### CODE REVIEW

File: tikv/raftstore/peer_storage.go [hunk 1]
```diff
@@ -291,6 +291,37 @@ func getMsg(engine *badger.DB, key []byte, msg proto.Message) error {
...
+
+func getRegionLocalStateMsg(db *badger.DB, regionId uint64) (*rspb.RegionLocalState, error) {
+	regionLocalState := new(rspb.RegionLocalState)
+	if err := getMsg(db, RegionStateKey(regionId), regionLocalState); err != nil {
+		return nil, &ErrRegionNotFound{regionId}
+	}
+	return regionLocalState, nil
+}
+
+func getRaftApplyStateMsg(db *badger.DB, regionId uint64) (*rspb.RaftApplyState, error) {
```
Comment (coocood) [inline tikv/raftstore/peer_storage.go]: `RaftApplyState` has been replaced by `raftState` in `state.go`.

File: tikv/raftstore/snap.go [hunk 2]
```diff
@@ -39,22 +39,21 @@ const (
 	snapshotVersion     = 2
 	deleteRetryMaxTime  = 6
 	deleteRetryDuration = 500 * time.Millisecond
-
-	JobStatusPending    uint64 = 0
-	JobStatusRunning    uint64 = 1
-	JobStatusCancelling uint64 = 2
-	JobStatusCancelled  uint64 = 3
-	JobStatusFinished   uint64 = 4
-	JobStatusFailed     uint64 = 5
```
Comment (coocood) [inline tikv/raftstore/snap.go]: Why remove them?

File: tikv/raftstore/peer_storage.go [hunk 3]
```diff
@@ -291,6 +291,37 @@ func getMsg(engine *badger.DB, key []byte, msg proto.Message) error {
 	return proto.Unmarshal(val, msg)
 }
 
+type storageError string
+
+func (e storageError) Error() string {
+	return string(e)
+}
+
+func getRegionLocalStateMsg(db *badger.DB, regionId uint64) (*rspb.RegionLocalState, error) {
```
Comment (coocood) [inline tikv/raftstore/peer_storage.go]: We can omit `Msg` for the function name.

File: tikv/raftstore/peer_storage.go [hunk 4]
```diff
@@ -291,6 +291,37 @@ func getMsg(engine *badger.DB, key []byte, msg proto.Message) error {
...
+func getRaftApplyStateMsg(db *badger.DB, regionId uint64) (*rspb.RaftApplyState, error) {
+	applyState := new(rspb.RaftApplyState)
+	err := getMsg(db, ApplyStateKey(regionId), applyState)
+	if err != nil {
+		return nil, storageError(fmt.Sprintf("couldn't load raft state of region %d", regionId))
+	}
+	return applyState, nil
+}
+
+func getRaftEntryMsg(db *badger.DB, regionId, idx uint64) (*eraftpb.Entry, error) {
```
Comment (coocood) [inline tikv/raftstore/peer_storage.go]: ditto

File: tikv/raftstore/worker.go [hunk 5]
```diff
@@ -3,7 +3,9 @@ package raftstore
 import (
 	"bytes"
 	"encoding/hex"
+	"fmt"
 	"github.com/coocood/badger/y"
+	"github.com/pingcap/kvproto/pkg/raft_serverpb"
```
Comment (coocood) [inline tikv/raftstore/worker.go]: We name `raft_serverpb` as `rspb` because the name is too long.

File: tikv/raftstore/worker.go [hunk 6]
```diff
@@ -338,31 +341,89 @@ func (snapCtx *snapContext) generateSnap(regionId uint64, notifier chan<- *eraft
 	return nil
 }
 
-type applySnapAbortError string
-
-func (e applySnapAbortError) Error() string {
-	return "abort"
+// clearUpOriginData clear up the region data before applying snapshot
```
Comment (coocood) [inline tikv/raftstore/worker.go]: cleanUp

File: tikv/raftstore/worker.go [hunk 7]
```diff
@@ -338,31 +341,89 @@ func (snapCtx *snapContext) generateSnap(regionId uint64, notifier chan<- *eraft
...
+	}
+
+	regionKey := RegionStateKey(regionId)
+	regionState, err := getRegionLocalStateMsg(snapCtx.engiens.kv.db, regionId)
+	if err != nil {
+		return errors.New(fmt.Sprintf("failed to get regionState from %v", regionKey))
+	}
+
+	// Clear up origin data
+	if err := snapCtx.clearUpOriginData(regionState, status); err != nil {
```
Comment (coocood) [inline tikv/raftstore/worker.go]: ditto

File: tikv/raftstore/peer_storage.go (hunk 1 above)
Comment (xiaobogaga) [reply tikv/raftstore/peer_storage.go]: yes

File: tikv/raftstore/peer_storage.go (hunk 3 above)
Comment (xiaobogaga) [reply tikv/raftstore/peer_storage.go]: yes

File: tikv/raftstore/peer_storage.go (hunk 4 above)
Comment (xiaobogaga) [reply tikv/raftstore/peer_storage.go]: yes

File: tikv/raftstore/worker.go (hunk 5 above)
Comment (xiaobogaga) [reply tikv/raftstore/worker.go]: ok

File: tikv/raftstore/worker.go (hunk 6 above)
Comment (xiaobogaga) [reply tikv/raftstore/worker.go]: yes,

File: tikv/raftstore/worker.go (hunk 7 above)
Comment (xiaobogaga) [reply tikv/raftstore/worker.go]: ok

**Comment by xiaobogaga** [comment]
@coocood comment are addressed and part test of regionWorker are ported. PTAL again thanks.

File: tikv/raftstore/engine.go [hunk 8]
```diff
@@ -225,3 +235,117 @@ func (wb *WriteBatch) Reset() {
...
+	// Delete lock
+	lockIte := db.lockStore.NewIterator()
+	keys = keys[:0]
+	keys = collectLockOrRollbackRangeKeys(lockIte, startKey, endKey, keys)
+	keys = collectLockOrRollbackRangeKeys(lockIte, oldStartKey, oldEndKey, keys)
+	if err := deleteLocksInBatch(db, keys, delRangeBatchSize); err != nil {
+		return err
+	}
+
+	// Delete rollback
```
Comment (coocood) [inline tikv/raftstore/engine.go]: We don't need to delete rollback, it will be GCed automatically.

File: tikv/raftstore/engine.go [hunk 9]
```diff
@@ -225,3 +235,117 @@ func (wb *WriteBatch) Reset() {
...
+	reader.Close()
+	if err := deleteKeysInBatch(db, keys, delRangeBatchSize); err != nil {
+		return err
+	}
+
+	// Delete lock
+	lockIte := db.lockStore.NewIterator()
+	keys = keys[:0]
+	keys = collectLockOrRollbackRangeKeys(lockIte, startKey, endKey, keys)
+	keys = collectLockOrRollbackRangeKeys(lockIte, oldStartKey, oldEndKey, keys)
```
Comment (coocood) [inline tikv/raftstore/engine.go]: We don't need to collect lock for old key range, there is no old key in lockstore.

File: tikv/raftstore/peer_storage.go [hunk 10]
```diff
@@ -297,24 +297,25 @@ func (e storageError) Error() string {
...
 	if err := getMsg(db, RegionStateKey(regionId), regionLocalState); err != nil {
 		return nil, &ErrRegionNotFound{regionId}
 	}
 	return regionLocalState, nil
 }
 
-func getRaftApplyStateMsg(db *badger.DB, regionId uint64) (*rspb.RaftApplyState, error) {
-	applyState := new(rspb.RaftApplyState)
-	err := getMsg(db, ApplyStateKey(regionId), applyState)
+func getApplyState(db *badger.DB, regionId uint64) (*applyState, error) {
```
Comment (coocood) [inline tikv/raftstore/peer_storage.go]: We can return struct type.

File: tikv/raftstore/engine.go (hunk 8 above)
Comment (xiaobogaga) [reply tikv/raftstore/engine.go]: ok

File: tikv/raftstore/engine.go (hunk 9 above)
Comment (xiaobogaga) [reply tikv/raftstore/engine.go]: ok

File: tikv/raftstore/peer_storage.go (hunk 10 above)
Comment (xiaobogaga) [reply tikv/raftstore/peer_storage.go]: ok

**Comment by xiaobogaga** [comment]
@coocood PTAL again, thanks.

File: tikv/raftstore/engine.go [hunk 11]
```diff
@@ -225,3 +235,107 @@ func (wb *WriteBatch) Reset() {
...
+			dbBatch.DeleteLock(key)
+		}
+		if err := dbBatch.WriteToKV(db); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func deleteRollbackInBatch(db *DBBundle, keys [][]byte, batchSize int) error {
```
Comment (coocood) [inline tikv/raftstore/engine.go]: This can be removed.

File: tikv/raftstore/engine.go [hunk 12]
```diff
@@ -95,6 +98,13 @@ func (wb *WriteBatch) DeleteLock(key []byte) {
 	})
 }
 
+func (wb *WriteBatch) DeleteRollback(key []byte) {
```
Comment (coocood) [inline tikv/raftstore/engine.go]: This can be removed.

File: tikv/raftstore/engine.go (hunk 11 above)
Comment (xiaobogaga) [reply tikv/raftstore/engine.go]: yes. thanks

File: tikv/raftstore/engine.go (hunk 12 above)
Comment (xiaobogaga) [reply tikv/raftstore/engine.go]: yes,

**Comment by xiaobogaga** [comment]
@coocood PTAL again, thanks.

**Pull request merged**